text/0806-engine-router-helpers.md

```diff
@@ -0,0 +1,273 @@
+---
+Stage: Accepted
+Start Date: 2022-03-11
+Release Date: Unreleased
+Release Versions:
+  ember-source: vX.Y.Z
+  ember-engines: vX.Y.Z
+Relevant Team(s): Ember.js
+RFC PR: https://github.com/emberjs/rfcs/pull/806
+---
+
+<!--- 
+Directions for above: 
+
+Stage: Leave as is
+Start Date: Fill in with today's date, YYYY-MM-DD
+Release Date: Leave as is
+Release Versions: Leave as is
+Relevant Team(s): Fill this in with the [team(s)](README.md#relevant-teams) to which this RFC applies
+RFC PR: Fill this in with the URL for the Proposal RFC PR
+-->
+
+# Engine Router Helpers
+
+## Summary
+
+This RFC intends to:
+
+- deprecate `<LinkToExternal />` component provided by `ember-engines` package.
+- introduce a new template helper `{{path-for}}` provided by new `ember-engines-router-helpers` package.
+
+## Motivation
+
+`<LinkToExternal />` is the primary way to link to a path that isn't below the mount point of the Ember Engine.
+
+`<LinkToExternal />` is a thin wrapper around `<LinkTo />` component and this worked well in Ember.js pre v4.
+Extension of `<LinkTo />` component has been deprecated in 3.27.0 version of Ember.js per [RFC #0671](./0671-modernize-built-in-components-1.md) and removed in 4.0.0 hence we need a new approach.
+
+There is another problem this RFC seeks to solve.
+Large scale application authors often create an addon(s) to contain reusable components  that can be rendered either in the context of the host app or Ember Engine.
+In such scenario an extra guard needs to be introduced as `<LinkToExternal />` component  available only in the context of Ember Engine and does not exist in the context of regular application, so it's common to see such pattern:
+
+```hbs
+{{#if @isEngine}}
+  <LinkToExternal @route="contact">Contact Us</LinkToExternal>
+{{else}}
+  <LinkTo @route="contact">Contact Us</LinkTo>
+{{/if}}
+```
+
+## Detailed design
+
+Changes proposed in scope of this RFC span several packages:
+
+* updates to `ember-engines`
+* updates to `ember-source`
+* introduction of new addon `ember-engines-router-helpers`
+
+### `ember-engines-router-helpers` addon
+
+This RFC proposes to introduce new addon `ember-engines-router-helpers` so the template helper `{{path-for}}` would be available for use within Ember Engine as well as in libraries with components that may be shared between conventional apps and Engines.
+
+Template helper `{{path-for}}` should detect the execution context and return different value if the owner is an instance of the `ApplicationInstance` or `EngineInstance`.
```

Comment:
While such a template helper makes for convenient usage in templates (and should be kept for that purpose), I think it's crucial that this functionality is also exposed on a service. Otherwise it will be impossible or needlessly complicated to invoke it from JS/TS.

The helper could then just delegate to that service.

Reply:
I agree with the sentiment that there should also be a service, I'll need to reflect a little on whether an engine specific version of router vs a service with a different name is better/worse. Thanks for the prior art links!

Reply:
@runspired @buschtoens as a prior art, there is https://github.com/ember-engines/ember-engines/issues/779, followed by PR https://github.com/ember-engines/ember-engines/pull/669.

@villander also released that functionality as separate addon https://github.com/villander/ember-engines-router-service

I do agree that it's essential to have to have such functionality exposed on a service which we can tackle independently from this RFC by pushing https://github.com/ember-engines/ember-engines/issues/779 and https://github.com/ember-engines/ember-engines/pull/669.